vulcan_geometry.py

```diff
@@ -0,0 +1,259 @@
+
+    # split the data into banks
+    for i, point_label in enumerate(positions['Point']):
+        bank_label = ''
+        if point_label.startswith('HA_'):
+            bank_label = 'bank5'
+        elif point_label.startswith('BR_'):
+            bank_label = 'bank1'
+        elif point_label.startswith('BL_'):
+            bank_label = 'bank2'
+        if bank_label:
+            banks_allpixels[bank_label]['Point'].append(point_label.split('_')[-1])
+            banks_allpixels[bank_label]['X'].append(positions['X'][i])
+            banks_allpixels[bank_label]['Y'].append(positions['Y'][i])
+            banks_allpixels[bank_label]['Z'].append(positions['Z'][i])
+
+    # convert the positions to ndarray of float
+    for bank_label in banks_allpixels.keys():
+        for column in ['X', 'Y', 'Z']:
+            banks_allpixels[bank_label][column] = np.array(banks_allpixels[bank_label][column], dtype=float)
+
+    def pointFromName(pixels, label: str) -> Vector:
+        index = pixels['Point'].index(label)
+        return Vector(pixels['X'][index], pixels['Y'][index], pixels['Z'][index])
+
+    def rectangleFromName(pixels, corners) -> Rectangle:
+        points = [pointFromName(pixels, name) for name in corners]
+        return Rectangle(*points, tolerance_len=0.035)
+
+    # get the 4 corners into a structure to return
+    # corners in the order LL, UL, UR, LR
+    # Rectangle says lower-left then clockwise
+    # survey/alignment labeled them as "tube 1" (i.e. T1) as most downstream
+    banks = {}
+    banks['bank1'] = rectangleFromName(banks_allpixels['bank1'],
+                                       ('D1T1B', 'D1T1T', 'D20T4T', 'D20T1B'))
+    banks['bank2'] = rectangleFromName(banks_allpixels['bank2'],
+                                       ('D20T1B', 'D20T4T', 'D1T1T', 'D1T1B'))
+    banks['bank5'] = rectangleFromName(banks_allpixels['bank5'],
+                                       ('D1T1B', 'D1T1T', 'D9T4T', 'D9T4B'))
+
+    return banks
+
+
+def addEightPack(instr, name: str, tube_type: str, upsidedown: bool = False):
+    '''Add an interleaved 8-pack where pixel 1 is in the lower-left
+    corner in the back then increases along the tube upward. The bottom
+    of tube two is in the front.
+
+    back     1 3 5 7
+    front     2 4 6 8
+
+    Setting ``upsidedown=True`` configures the pixels as
+
+    back      1 3 5 7
+    front    2 4 6 8
+
+    This is similar to the incomplete function MantidGeometry.add_double_pack.
+    '''
+    type_element = le.SubElement(instr.root, 'type', name=name)
+    le.SubElement(type_element, 'properties')
+    component = le.SubElement(type_element, 'component', type=tube_type)
+    # z plane is centered in middle of detectors
+    tube_z = (-1. * SEPARATION, SEPARATION)  # 2 rows back, front
+    # x plane is centered between the leftmost and rightmost tubes
+    if upsidedown:
+        x_offset = (-1.25 * SLIP, -1.75 * SLIP)
+    else:
+        x_offset = (-1.75 * SLIP, -1.25 * SLIP)
```

Comment:
Can we add a comment to state how the two number, 1.25 and 1.75, were determined? It will help future developer understand a little bit better how x_offset was defined.